File: AFMobile/AFAndroid/app/src/main/java/cz/cvut/fel/matyapav/afandroid/builders/widgets/OptionWidgetBuilder.java
package cz.cvut.fel.matyapav.afandroid.builders.widgets;

import android.app.Activity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.RadioButton;
import android.widget.RadioGroup;

import com.tomscz.afrest.rest.dto.AFFieldInfo;

import cz.cvut.fel.matyapav.afandroid.components.parts.AFField;
import cz.cvut.fel.matyapav.afandroid.builders.skins.Skin;
import cz.cvut.fel.matyapav.afandroid.utils.Localization;

/**
 * Builds radio group widget, where user can choose from multiple options.
 *
 * @author Pavel Matyáš (matyapav@fel.cvut.cz)
 *
 * @since 1.0.0.
 */
public class OptionWidgetBuilder extends BasicWidgetBuilder {

    public OptionWidgetBuilder(Skin skin, AFField field) {
        super(skin, field);
    }

    @Override
    public View buildFieldView(Activity activity) {
        RadioGroup radioGroup = new RadioGroup(activity);
        radioGroup.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
        radioGroup.setOrientation(RadioGroup.VERTICAL); //TODO change according to layout parameters
        int numberOfOptions;
        if(getField().getFieldInfo().getOptions() != null && !getField().getFieldInfo().getOptions().isEmpty()) {
            numberOfOptions = getField().getFieldInfo().getOptions().size();
            RadioButton[] options = new RadioButton[numberOfOptions];
            for (int i = 0; i < numberOfOptions; i++) {
                options[i] = new RadioButton(activity);
                options[i].setTextColor(getSkin().getFieldColor());
                options[i].setTypeface(getSkin().getFieldFont());
                options[i].setText(getField().getFieldInfo().getOptions().get(i).getValue());
                options[i].setId(i + 100);
                radioGroup.addView(options[i]);
            }
        }else{
            //YES NO OPTIONS
            numberOfOptions = 2;
            RadioButton[] options = new RadioButton[numberOfOptions];
            options[0] = new RadioButton(activity);
            options[0].setTextColor(getSkin().getFieldColor());
            options[0].setTypeface(getSkin().getFieldFont());
            options[0].setText(Localization.translate("option.yes"));
            options[1] = new RadioButton(activity);
            options[1].setTextColor(getSkin().getFieldColor());
            options[1].setTypeface(getSkin().getFieldFont());
            options[1].setText(Localization.translate("option.no"));
            radioGroup.addView(options[0]);
            radioGroup.addView(options[1]);
        }
        if(getField().getFieldInfo().getReadOnly()){
            radioGroup.setEnabled(false);
        }
        return radioGroup;
    }

    //TODO REWRITE
    @Override
    public void setData( Object value) {

        RadioGroup group = (RadioGroup) getField().getFieldView();
        if(value == null){
            group.clearCheck();
            getField().setActualData(value);
            return;
        }
        for (int i = 0; i < group.getChildCount(); i++) {
            RadioButton btn = (RadioButton) group.getChildAt(i);
            if(btn.getText().equals(value)
                    || (Boolean.valueOf(value.toString()) == true && i==0)
                    || (Boolean.valueOf(value.toString()) == false && i==1)){
                btn.setChecked(true);
                getField().setActualData(value);
                break;
            }
        };

    }

    @Override
    public Object getData() {
        RadioGroup group = (RadioGroup)  getField().getFieldView();
        for (int i = 0; i < group.getChildCount(); i++) {
            RadioButton btn = (RadioButton) group.getChildAt(i);
            if(btn.isChecked()){
                if(btn.getText().toString().equals(Localization.translate("option.yes"))){
                    return true;
                }else if(btn.getText().toString().equals(Localization.translate("option.no"))){
                    return false;
                }else {
                    return btn.getText().toString();
                }
            }
        }
        return null; //nothing is checked
    }
}

File: AFMobile/AFWinPhone/AFWinPhone/builders/widgets/OptionWidgetBuilder.cs
﻿using AFWinPhone.components.parts;
using AFWinPhone.builders.skins;
using System;
using Windows.UI.Xaml;
using System.Linq;
using Windows.UI.Xaml.Controls;
using AFWinPhone.utils;

namespace AFWinPhone.builders.widgets
{
    /// <summary>
    /// Builds radio group widget, where user can choose from multiple options.
    /// </summary>
    class OptionWidgetBuilder : BasicBuilder
    {
        public OptionWidgetBuilder(Skin skin, AFField field) : base(skin, field)
        {
        }

        public override FrameworkElement buildFieldView()
        {
            StackPanel radioGroup = new StackPanel();
            radioGroup.HorizontalAlignment = HorizontalAlignment.Stretch;
            radioGroup.VerticalAlignment = VerticalAlignment.Top;
            radioGroup.Orientation = Orientation.Vertical;
            int numberOfOptions;
            if (getField().getFieldInfo().getOptions() != null && getField().getFieldInfo().getOptions().Any())
            {
                numberOfOptions = getField().getFieldInfo().getOptions().Count;
                RadioButton[] options = new RadioButton[numberOfOptions];
                String groupName = Guid.NewGuid().ToString();
                for (int i = 0; i < numberOfOptions; i++)
                {
                    options[i] = new RadioButton();
                    options[i].FontFamily = getSkin().getFieldFont();
                    options[i].FontSize = getSkin().getFieldFontSize();
                    options[i].Content = getField().getFieldInfo().getOptions()[i].getValue();
                    options[i].GroupName = groupName;
                    if (getField().getFieldInfo().isReadOnly())
                    {
                        options[i].IsEnabled = false;
                    }
                    radioGroup.Children.Add(options[i]);
                }
            }
            else {
                //YES NO OPTIONS
                numberOfOptions = 2;
                RadioButton[] options = new RadioButton[numberOfOptions];
                options[0] = new RadioButton();
                options[0].FontFamily = getSkin().getFieldFont();
                options[0].FontSize = getSkin().getFieldFontSize();
                options[0].Content = Localization.translate("option.yes");
                options[1] = new RadioButton();
                options[1].FontFamily = getSkin().getFieldFont();
                options[1].FontSize = getSkin().getFieldFontSize();
                options[1].Content = Localization.translate("option.no");
                if (getField().getFieldInfo().isReadOnly())
                {
                    options[0].IsEnabled = false;
                    options[1].IsEnabled = false;
                }
                radioGroup.Children.Add(options[0]);
                radioGroup.Children.Add(options[1]);
            }
          
            return radioGroup;
        }

        public override void setData(Object value)
        {

            StackPanel group = (StackPanel) getField().getFieldView();
            if (value == null)
            {
                foreach (RadioButton btn in group.Children)
                {
                    btn.IsChecked = false;
                }
                getField().setActualData(value);
                return;
            }
            for (int i = 0; i < group.Children.Count; i++)
            {
                RadioButton btn = (RadioButton) group.Children[i];
                if (btn.Content.Equals(value)
                        || (Utils.TryToConvertIntoBoolean(value.ToString()) == true && i == 0)
                        || (Utils.TryToConvertIntoBoolean(value.ToString()) == false && i == 1))
                {
                    btn.IsChecked = true;
                    getField().setActualData(value);
                    break;
                }
            };

        }

        public override Object getData()
        {
            StackPanel group = (StackPanel) getField().getFieldView();
      
            for (int i = 0; i < group.Children.Count; i++)
            {
                RadioButton btn = (RadioButton) group.Children[i];
                if (btn.IsChecked.Value)
                {
                    if (btn.Content.Equals(Localization.translate("option.yes")))
                    {
                        return "true";
                    }
                    else if (btn.Content.Equals(Localization.translate("option.no")))
                    {
                        return "false";
                    }
                    else {
                        return btn.Content;
                    }
                }
            }
            return null; //nothing is checked
        }
    }
}
